### PR TITLE
refactor: source entity to add flags property

### PR DESCRIPTION
**Pull request by sshanzel**
Introduce a new `flags` column for `source` entity. Workers and such will be worked on a different PR.

MI-321 #done

### CODE REVIEW

File: src/entity/Source.ts
Comment (capJavert) [inline src/entity/Source.ts:36]: Flags should be empty by default, so any defaulting to zero should be handled in other API logic or on graphorm side. 

For example having defaultValues and then updating them after some time would require migration to update all previous rows that had different default value. Which is complicated to do.

File: src/entity/Source.ts
Comment (capJavert) [inline src/entity/Source.ts:108]: @sshanzel since this is such a small PR can you just add "real logic" like workers in it as well? Just to make sure we did not miss something here and then require additional migrations in follow up PRs. If you want to chunk it still let's make an epic branch for this series of tasks?

Comment (sshanzel) [reply src/entity/Source.ts:108]: I agree making it be merged all at once would allow us to have a better view on this whole series of changes. Though I think part of the reason why the tasks were chunked into smaller ones is to allow working in parallel. From the JIRA ticket, this blocks 3 items, merging this small PR early would allow us to work on those 3 tickets at the same time.

Yeah, maybe we can create an epic branch here in the API.

File: src/entity/Source.ts
Comment (sshanzel) [reply src/entity/Source.ts:36]: I actually did that at first but hesitated so I went to the RFC and checked if we specified a default value, found they mentioned 0 so I went ahead and applied it. @DragosIorgulescu thoughts?

File: src/entity/Source.ts
Comment (capJavert) [reply src/entity/Source.ts:108]: @sshanzel yeah got it, if you can at least hook up graphorm in the PR and at tests for the field and I am fine with merging this as well. You can see example on `flags field` tests in `posts`.

File: src/entity/Source.ts
Comment (capJavert) [inline src/entity/Source.ts:30]: Also please follow `SourceFlagsPublic` nomenclature, you can see example for  `UserPostEntity`. Basically public flags are those that can be requested/updated from graphql.

File: src/entity/Source.ts
Comment (sshanzel) [reply src/entity/Source.ts:108]: Yeah, was thinking that would allow us to unblock FE work too. Sure, let me update those items.

File: src/entity/Source.ts
Comment (rebelchris) [reply src/entity/Source.ts:36]: It can be solved on graphORM side not db side basically.

File: src/migration/1714116154573-SourceFlags.ts
Comment (rebelchris) [inline src/migration/1714116154573-SourceFlags.ts:8]: We should remove the defaults before merging :)

File: src/entity/Source.ts
Comment (sshanzel) [reply src/entity/Source.ts:36]: Yeah, I understand, it is just on the RFC, the table was dedicated to the entity, not the gql schema. In any case, I will revert to empty curly brackets similar to our other implementations of flags.

**Review by rebelchris — APPROVED** [review]
Looks good to me, just wait for ante 👍

File: src/schema/sources.ts
Comment (capJavert) [inline src/schema/sources.ts:1652]: Also I think better to do it on graphorm level because it will be available directly in queries vs only after query resolves. Basically under `jsonType: true` with `transform`.

File: src/graphorm/index.ts
Comment (capJavert) [inline src/graphorm/index.ts:228]:
```suggestion
        jsonType: true,
        transform: (flags: Partial<SourceFlagsPublic>) => {
          return {
            totalPosts: flags?.totalPosts ?? 0,
            totalViews: flags?.totalViews ?? 0,
            totalUpvotes: flags?.totalUpvotes ?? 0,
            featured: flags?.featured ?? false,
          };
        },
```

Comment (capJavert) [reply src/graphorm/index.ts:228]: Also if you can add tests, check here for example 🙏 https://github.com/dailydotdev/daily-api/blob/main/__tests__/sources.ts#L2795-L2834

**Review by capJavert — APPROVED** [review]
Looks good, just some minor adjustments.

File: src/schema/sources.ts
Comment (capJavert) [reply src/schema/sources.ts:1652]: https://github.com/dailydotdev/daily-api/pull/1857#discussion_r1581177349

Comment (sshanzel) [reply src/schema/sources.ts:1652]: Yeah, that makes sense, we will have more control and avoid the null value on certain situations.

Comment (sshanzel) [reply src/schema/sources.ts]: Introduced a function to determine the value so it will be consistent.

File: src/graphorm/index.ts
Comment (sshanzel) [reply src/graphorm/index.ts:231]: I introduced a new function as I realized the properties can grow and the scenario can easily happen in the future. That is to make the process consistent.

File: __tests__/commons/object.ts
Comment (nensidosari) [inline __tests__/commons/object.ts:17]:
```suggestion
  it('should set the default value if property is null or undefined', () => {
```

File: __tests__/commons/object.ts
Comment (nensidosari) [inline __tests__/commons/object.ts:26]:
```suggestion
  it('should NOT set the default value if property is NOT null or undefined', () => {
```

**Review by nensidosari — APPROVED** [review]
Looks good on my side as well

**Review by DragosIorgulescu — APPROVED** [review]
nothing to add 🙌

File: src/common/object.ts
Comment (capJavert) [inline src/common/object.ts:63]: While I get what you are trying to do this seems too complicated in my opinion. 
- it is doing iteration (over keys) which just adds more computation on field resolver using it
- what if someone introduces nested object, this would not be able to resolve it? 

I personally much more prefer just doing it inline in gql as you did before https://github.com/dailydotdev/daily-api/pull/1857/commits/e31fd547bf5bc65c2e4e4b152a8e6295860d2e5c#diff-95123b05e21b34d3705f1b3748fe295fa17db5caeeb22238a294368729175a07R1649-R1652

Comment (capJavert) [reply src/common/object.ts:63]: @rebelchris @DragosIorgulescu tagging for your opinion as well.

Comment (sshanzel) [reply src/common/object.ts:63]: Yeah, I just realized if we add another property, we might miss another `obj.key ?? value` to put since everything is inline. For now, yeah, this won't cover a nested object, but that can definitely add that processing once we get the need for it.

Comment (capJavert) [reply src/common/object.ts:63]: > Yeah, I just realized if we add another property, we might miss another obj.key ?? value 

Yeah, figured, did you look into option to take care of this with Typescript instead by defining return type of field resolver, which would be caught in lint/test before any deployment. So it would make sure additional fields do not miss handling without introducing runtime cost.

Comment (sshanzel) [reply src/common/object.ts:63]: Yeah, I checked, and it could be possible, just needs more work on the GraphORM side since it doesn't automatically understand the type it needs to return.

Comment (capJavert) [reply src/common/object.ts:63]: I tested with:
```ts
transform: (value: SourceFlagsPublic): SourceFlagsPublic => {
  return {
    totalPosts: value?.totalPosts ?? 0,
    totalViews: value?.totalViews ?? 0,
    totalUpvotes: value?.totalUpvotes ?? 0,
    featured: value?.featured ?? false,
  };
}
```

and then adding new field `test` to `SourceFlagsPublic`

I automatically get the error:
image

Which would block any pipeline and force developer to add handling in graphorm.

Comment (sshanzel) [reply src/common/object.ts:63]: Right. Yeah, we can probably do that instead of adding generic types to GraphORM.

Comment (sshanzel) [reply src/common/object.ts:63]: Ideally, we can do generic types for better devex and avoid forgetting to add that type on the function since our lint does not ask for it.

Comment (capJavert) [reply src/common/object.ts]: Yeah, problem partially is that we don't have [strict option in typescript](https://dailydotdev.atlassian.net/browse/WT-1532), it would light up red all over the place even more during dev.

**Pull request merged**
